File: src/voice-agent/sidecar/ow_conversation/protocol.py
"""Mirror of src/voice-agent/shared/protocol.mjs for the Python side of the loopback WebSocket."""

from __future__ import annotations

import json
import time
from typing import Any

PROTOCOL_VERSION = 1
MAX_MESSAGE_BYTES = 1024 * 1024

SIDECAR_MESSAGE_TYPES = frozenset(
    {
        "hello",
        "ready",
        "state",
        "transcript.partial",
        "transcript.final",
        "assistant.delta",
        "assistant.final",
        "tool.call",
        "tool.cancel",
        "latency",
        "level",
        "warning",
        "error",
        "bye",
    }
)

APP_MESSAGE_TYPES = frozenset(
    {"session.config", "tool.result", "say", "interrupt", "mute", "update_tools", "shutdown", "ping"}
)


class ProtocolError(ValueError):
    pass


def encode(message_type: str, data: dict[str, Any] | None = None, *, id: str | None = None, turn: int | None = None) -> str:
    if message_type not in SIDECAR_MESSAGE_TYPES:
        raise ProtocolError(f"cannot send {message_type!r}")
    message: dict[str, Any] = {
        "v": PROTOCOL_VERSION,
        "type": message_type,
        "ts": int(time.time() * 1000),
        "data": data or {},
    }
    if id is not None:
        message["id"] = id
    if turn is not None:
        message["turn"] = turn
    return json.dumps(message, ensure_ascii=False)


def decode(raw: str | bytes) -> dict[str, Any]:
    if isinstance(raw, bytes):
        raise ProtocolError("binary frames are not accepted")
    if len(raw) > MAX_MESSAGE_BYTES:
        raise ProtocolError("too-large")
    try:
        message = json.loads(raw)
    except json.JSONDecodeError as exc:
        raise ProtocolError("invalid-json") from exc
    if not isinstance(message, dict):
        raise ProtocolError("not-object")
    if message.get("v") != PROTOCOL_VERSION:
        raise ProtocolError("version")
    if message.get("type") not in APP_MESSAGE_TYPES:
        raise ProtocolError("type")
    data = message.get("data", {})
    if not isinstance(data, dict):
        raise ProtocolError("data")
    if "id" in message and not isinstance(message["id"], str):
        raise ProtocolError("id")
    message["data"] = data
    return message

File: src/voice-agent/shared/protocol.mjs
// Loopback WebSocket protocol between Electron (server) and the Python voice sidecar (client).
// Pure logic. Every message is a JSON object: { v, type, ts, id?, turn?, data }.

export const PROTOCOL_VERSION = 1;
export const MAX_MESSAGE_BYTES = 1024 * 1024;

export const SIDECAR_MESSAGE_TYPES = Object.freeze([
  "hello",
  "ready",
  "state",
  "transcript.partial",
  "transcript.final",
  "assistant.delta",
  "assistant.final",
  "tool.call",
  "tool.cancel",
  "latency",
  "level",
  "warning",
  "error",
  "bye",
]);

export const APP_MESSAGE_TYPES = Object.freeze([
  "session.config",
  "tool.result",
  "say",
  "interrupt",
  "mute",
  "update_tools",
  "shutdown",
  "ping",
]);

// What may reach renderer windows. tool.call goes only to the tool executor, and nothing the
// sidecar sends is relayed unless listed here.
export const RELAYED_TO_WINDOWS = Object.freeze([
  "state",
  "transcript.partial",
  "transcript.final",
  "assistant.delta",
  "assistant.final",
  "latency",
  "level",
  "warning",
  "error",
]);

export function encodeMessage(type, data = {}, extra = {}) {
  const message = { v: PROTOCOL_VERSION, type, ts: Date.now(), data };
  if (typeof extra.id === "string") message.id = extra.id;
  if (typeof extra.turn === "number") message.turn = extra.turn;
  return JSON.stringify(message);
}

/**
 * @param {string} raw
 * @param {readonly string[]} allowedTypes
 * @returns {{ ok: true, message: object } | { ok: false, error: string }}
 */
export function decodeMessage(raw, allowedTypes) {
  if (typeof raw !== "string") return { ok: false, error: "not-text" };
  if (raw.length > MAX_MESSAGE_BYTES) return { ok: false, error: "too-large" };
  let message;
  try {
    message = JSON.parse(raw);
  } catch {
    return { ok: false, error: "invalid-json" };
  }
  if (!message || typeof message !== "object" || Array.isArray(message)) {
    return { ok: false, error: "not-object" };
  }
  if (message.v !== PROTOCOL_VERSION) return { ok: false, error: "version" };
  if (typeof message.type !== "string" || !allowedTypes.includes(message.type)) {
    return { ok: false, error: "type" };
  }
  if (
    message.data !== undefined &&
    (typeof message.data !== "object" || message.data === null || Array.isArray(message.data))
  ) {
    return { ok: false, error: "data" };
  }
  if (message.id !== undefined && typeof message.id !== "string") return { ok: false, error: "id" };
  return { ok: true, message: { ...message, data: message.data || {} } };
}
